presentation/components/src/presentation-components/unified-selection/UnifiedSelectionContext.tsx

```diff
@@ -0,0 +1,145 @@
+*--------------------------------------------------------------------------------------------*/
+/**
+ * @packageDocumentation
+ * @module UnifiedSelection
+ */
+
+import * as React from "react";
+import { assert } from "@bentley/bentleyjs-core";
+import { IModelConnection } from "@bentley/imodeljs-frontend";
+import { Keys, KeySet } from "@bentley/presentation-common";
+import { Presentation } from "@bentley/presentation-frontend";
+
+/**
+ * Interface for accessing and modifying Unified Selection whithin React components.
+ * @public
+ */
+export interface UnifiedSelectionContext {
+  /** iModel associated with the selection. */
+  imodel: IModelConnection;
+
+  /** Default selection level. */
+  selectionLevel: number;
+
+  /** Returns current selection. */
+  getSelection(level?: number): Readonly<KeySet>;
+
+  /** Adds EC instances to current selection. */
+  addToSelection(keys: Keys, level?: number): void;
+
+  /** Removes EC instances from current selection. */
+  removeFromSelection(keys: Keys, level?: number): void;
+
+  /** Removes all EC instances from current selection and selects only the specified EC instances. */
+  replaceSelection(keys: Keys, level?: number): void;
+
+  /** Removes all EC instances from current selection. */
+  clearSelection(level?: number): void;
+}
+
+/**
+ * Props for Unified Selection context provider.
+ * @public
+ */
+export interface UnifiedSelectionContextProviderProps {
+  /** iModel associated with the selection. */
+  imodel: IModelConnection;
+
+  /**
+   * Specifies the selection level to watch for selection changes. This also becomes the default selection level for
+   * [[UnifiedSelectionContext]]. Default value is 0.
+   */
+  selectionLevel?: number;
+
+  children?: React.ReactNode;
+}
+
+/**
+ * Unified Selection context provider. It adapts framework-agnostic
+ * [Unified Selection]($docs/learning/presentation/Unified-Selection/index.md) API to be better suited for React
+ * applications. The context is accessible via [[useUnifiedSelectionContext]].
+ * @public
+ */
+export const UnifiedSelectionContextProvider: React.FC<UnifiedSelectionContextProviderProps> = (props) => {
+  const selectionLevel = props.selectionLevel ?? 0;
+
+  const contextRef = React.useRef<UnifiedSelectionContext>();
+  if (contextRef.current?.imodel !== props.imodel || contextRef.current.selectionLevel !== selectionLevel) {
+    contextRef.current = createContext(props.imodel, selectionLevel);
+  }
+
+  const [_, setState] = React.useState({});
+  React.useEffect(
+    () => {
+      const currentContext = contextRef.current;
+      assert(currentContext !== undefined);
+      return Presentation.selection.selectionChange.addListener((args) => {
+        if (args.level > currentContext.selectionLevel) {
+          return;
+        }
+
+        contextRef.current = {
+          ...currentContext,
+          getSelection: createGetSelection(props.imodel, selectionLevel),

```

Comment:
I don't think we need to re-create the `getSelection` function when selection changes. Consumers can use `KeySet.guid` as a dependency to track selection changes:
```ts
const context = useUnifiedSelectionContext();
const selection = context.getSelection();
const foo = React.useMemo(() => {
  // use `selection`
}, [selection.guid]);
```

Reply:
This change request didn't sit well with me, so I talked to @grigasp privately to figure out what the actual problem was.

What I uncovered, unfortunately, is even more questionable. The actual request is to make the returned reference to mutable `KeySet` be different on each selection change, but only keep one copy of the data, as it can be prohibitively expensive to copy. The intention is to let other components accept resulting `KeySet` as props, and to make them rerender when `KeySet` reference changes, despite containing the same mutable data structures. This will inevitably lead to some developers getting confused by their props getting spontaneously mutated and some hard to figure out bugs, but this idea was backed by multiple maintainers when asked for feedback.